### ISSUE TITLE
parallel processing and betweenness centrality?

### ISSUE THREAD

**Issue by evebohnett** [igraph/rigraph#315]
For really long jobs, this algorithm takes a long time to run. One of my jobs has been running for nearly 30 days. I wanted to use parallel computing and was wondering if it would be possible to edit the code for this betweenness function to run in a parallel environment? Has this been done already?

**Comment by alfonsocheng1025** [comment]
> For really long jobs, this algorithm takes a long time to run. One of my jobs has been running for nearly 30 days. I wanted to use parallel computing and was wondering if it would be possible to edit the code for this betweenness function to run in a parallel environment? Has this been done already?

Hi there!  Have you had this parallel computing problem sloved? I'm also doing the betweenness centrality calculations for 20000+ nodes with 4 billion+ edges amongst them.  I'm eager to use parallel computing to acceralate such process since it really takes me a long long time... Thx!

**Comment by szhorvat** [comment]
I am going to close this now as if this were implemented, it would need to be in igraph's C core, not specifically in the R interface. Feel free to open a feature request there, but be aware that this feature is not planned and would likely not be prioritized.

Some tips:

 - Before committing to a multi-hour or even multi-day computation, estimate how long it should actually take by running the program on increasingly larger instances. It makes no sense to wait for 30 days and just hope that it finishes unless you have good reason to believe that it will finish in that time. Parallelization would at best result in a speedup of a small constant factor, which is no help if the computation would take years.

 - There is an `estimate_betweenness` function, but it's name is misleading. What it actually does is compute a range-limited betweenness (see docs). This _can_ in fact be used to estimate the actual betweenness, see https://journals.aps.org/prl/abstract/10.1103/PhysRevLett.105.038701

**Comment by evebohnett** [comment]
There is a way to run betweens centrality in parallel in python. Although I was using raster data so finally were unable to complete this task because of the data structure and some of the advance level programming tasks. I had some of the guys here at the supercomputing center help me with some of the programming which fibally was not completed unfortunately. 
This would be a very useful tool if it could be run in parallel for landscape ecology applications. 

Eve Bohnett



 

    On Sunday, November 22, 2020, 03:31:32 AM PST, Szabolcs Horvát <notifications@github.com> wrote:  
 
 


I am going to close this now as if this were implemented, it would need to be in igraph's C core, not specifically in the R interface. Feel free to open a feature request there, but be aware that this feature is not planned and would likely not be prioritized.

Some tips:
   
   -    
Before committing to a multi-hour or even multi-day computation, estimate how long it should actually take by running the program on increasingly larger instances. It makes no sense to wait for 30 days and just hope that it finishes unless you have good reason to believe that it will finish in that time. Parallelization would at best result in a speedup of a small constant factor, which is no help if the computation would take years.

   -    
There is an estimate_betweenness function, but it's name is misleading. What it actually does is compute a range-limited betweenness (see docs). This can in fact be used to estimate the actual betweenness, see https://journals.aps.org/prl/abstract/10.1103/PhysRevLett.105.038701

**Comment by szhorvat** [comment]
> This would be a very useful tool if it could be run in parallel for landscape ecology applications. 

There is no doubt that it would be useful. As I said, feel free to open a feature request in the C/igraph repo: https://github.com/igraph/igraph/issues  The problem is that we are short on resources. _If anyone wants to put in the effort to implement this, contributions are always very welcome!_

> There is a way to run betweens centrality in parallel in python.

Just noting for other readers that this comment does not refer to the python interface of igraph. @evebohnett – I assume you used a different library.

**Comment by evebohnett** [comment]
Yes. I am actually it was networkx. They have a parallel betweenness function. 
Best,

Eve Bohnett



 

    On Sunday, November 22, 2020, 09:54:35 AM PST, Szabolcs Horvát <notifications@github.com> wrote:  
 
 



This would be a very useful tool if it could be run in parallel for landscape ecology applications.


There is no doubt that it would be useful. As I said, feel free to open a feature request in the C/igraph repo: https://github.com/igraph/igraph/issues The problem is that we are short on resources. If anyone wants to put in the effort to implement this, contributions are always very welcome!


There is a way to run betweens centrality in parallel in python.


Just noting for other readers that this comment does not refer to the python interface of igraph. @evebohnett – I assume you used a different library.

**Comment by szhorvat** [comment]
> Yes. I am actually it was networkx. They have a parallel betweenness function.

This is not quite correct. They have a function that computes the betweenness only for a subset of vertices. One can use this to write one's own parallelized version. 

igraph can also compute the betweenness for a subset of vertices, however, at the moment, subsetting happens only after the full computation was done, therefore it does not result in performance improvements.

It is good to note that networkx is significantly slower than igraph. On my machine, a betweenness in a 40x40 grid graph is computed in 10 seconds by networkx vs 0.2 seconds by igraph. For 60x60 the timings are 52 s vs 1 s. That's a 50-fold difference. Given the unavoidable overhead of parallelization, this suggests that one would need to run the parallel networkx version of a ~100-core machine to merely achieve parity with igraph. This is of course expected, as networkx is written in Python, while igraph in C.

This example illustrates well why "just parallelize it" (which is far from a trivial undertaking, if done right) is not usually the right solution to performance problems.

**Comment by evebohnett** [comment]
Okay great, I've added a feature request 
https://github.com/igraph/igraph/issues/1552

Maybe someone will feel inspired by the topics. 
I also noticed that szhorvat submitted as well. I guess these could be merged...my request is much less technical. 
Best,
Eve Bohnett



 

    On Sunday, November 22, 2020, 01:28:36 PM EST, Szabolcs Horvát <notifications@github.com> wrote:  
 
 



Yes. I am actually it was networkx. They have a parallel betweenness function.


This is not quite correct. They have a function that computes the betweenness only for a subset of vertices. One can use this to write one's own parallelized version.

igraph can also compute the betweenness for a subset of vertices, however, at the moment, subsetting happens only after the full computation was done, therefore it does not result in performance improvements.

It is good to note that networkx is significantly slower than igraph. On my machine, a betweenness in a 40x40 grid graph is computed in 10 seconds by networkx vs 0.2 seconds by igraph. For 60x60 the timings are 52 s vs 1 s. That's a 50-fold difference. Given the unavoidable overhead of parallelization, this suggests that one would need to run the parallel networkx version of a ~100-core machine to merely achieve parity with igraph. This is of course expected, as networkx is written in Python, while igraph in C.

This example illustrates well why "just parallelize it" (which is far from a trivial undertaking, if done right) is not usually the right solution to performance problems.
